### PR TITLE
Add-xmake-build-support

### PR DESCRIPTION
**Pull request by Arteiimis**

### CODE REVIEW

File: code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp
Comment (rsp4jack) [inline code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp:1]: 既然是 to add xmake build support，为何有诸如此类的改动？

Comment (rsp4jack) [inline code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp:1]: 既然是 to add xmake build support，为何有诸如此类的改动？

Comment (Arteiimis) [reply code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp:1]: 这里怀疑是白老师失误了，不改直接过不了编译

Comment (Mq-b) [reply code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp:1]: 我的提交日志中写的是：“视频代码”。
主要在于这是视频中写的，可能是一开始没写这些的时候会编译，然后写点代码，看看智能感知，也懒得全部写完，就放在那了，无法通过编译是正常情况，没什么奇怪的。
至于添加头文件则是合理做法，因为 msvc 可能就不小心有间接包含了，忘记引入。

File: code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #include <memory>
 
 std::unique_lock<std::mutex> get_lock() {
-    extern std::mutex some_mutex;
+    static std::mutex some_mutex;
```
Comment (Mq-b) [inline code/ModernCpp-ConcurrentProgramming-Tutorial/17-在不同作用域传递互斥量.cpp:8]: 此处实乃特意而为，视频与教案均如此，展示了许多做法。
非是简单改为 static 就可以的，虽然其的确也是视频中提到的一种方式。

**Comment by Mq-b** [comment]
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial> git fetch origin pull/30/head:pr-30
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 13 (delta 8), reused 10 (delta 8), pack-reused 0 (from 0)
Unpacking objects: 100% (13/13), 2.32 KiB | 113.00 KiB/s, done.
From https://github.com/Mq-b/ModernCpp-ConcurrentProgramming-Tutorial
 * [new ref]         refs/pull/30/head -> pr-30
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial> git checkout pr-30
M       "md/06\345\215\217\347\250\213.md"
Switched to branch 'pr-30'
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial> cd code
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial\code> xmake
note: xmake.lua not found, try generating it (pass -y or --confirm=y/n/d to skip confirm)?
please input: n (y/n)
n
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial\code> cd .\ModernCpp-ConcurrentProgramming-Tutorial\
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial\code\ModernCpp-ConcurrentProgramming-Tutorial> xmake
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022
checking for Microsoft C/C++ Compiler (x64) version ... 19.41.34120
updating repositories .. ok
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> freetype 2.13.1 [runtimes:"MD", from:sfml]
  -> libogg v1.3.4 [runtimes:"MD", from:libflac,sfml,libvorbis]
  -> libflac 1.4.3 [runtimes:"MD", from:sfml]
  -> libvorbis 1.3.7 [runtimes:"MD", from:sfml]
  -> openal-soft 1.23.1 [runtimes:"MD", from:sfml]
  -> sfml 2.6.0 [runtimes:"MD"]
  -> fmt 11.0.2 [runtimes:"MD"]
  -> aqt 3.1.17 [from:qt6base]
  -> qt6base 6.8.0 [from:qt6core]
  -> qt6core 6.8.0 [runtimes:"MD"]
  -> boost 1.86.0 [runtimes:"MD"]
please input: y (y/n/m)
y
  => download https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.4/ogg-v1.3.4.tar.gz .. ok
  => download https://github.com/xmake-mirror/build-artifacts/releases/download/fmt-11.0.2/windows-x64-vc14.40.33807-56f2f78a406c4e1aaac1ce081863e86e.7z .. ok
  => install fmt 11.0.2 .. ok
  => install libogg v1.3.4 .. ok
  => download https://github.com/xiph/flac/archive/1.4.3.tar.gz .. ok
  => install libflac 1.4.3 .. ok
  => download https://github.com/xmake-mirror/build-artifacts/releases/download/openal-soft-1.23.1/windows-x64-vc14.40.33807-efad308ac28c4beb9bf7be5247a85f61.7z .. ok
  => install openal-soft 1.23.1 .. ok
  => download https://github.com/xiph/vorbis/releases/download/v1.3.7/libvorbis-1.3.7.tar.gz .. ok
  => install libvorbis 1.3.7 .. ok
  => download https://downloads.sourceforge.net/project/freetype/freetype2/2.13.1/freetype-2.13.1.tar.gz .. failed     
  => download https://download.savannah.gnu.org/releases/freetype/freetype-2.13.1.tar.gz .. ok
  => install freetype 2.13.1 .. failed

生成启动时间为 2024/10/4 12:09:12。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”在节点 1 上(Rebuild 个目标)。
     1>CoreClean:
         正在创建目录“x64\Release\ALL_BUILD\”。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ZERO_CHECK.vcxproj”(2) (Clean 个目标)。
     2>CoreClean:
         正在创建目录“x64\Release\ZERO_CHECK\”。
     2>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ZERO_CHECK.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 2 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(3) (Clean 个目标)。
     3>CoreClean:
         正在创建目录“freetype.dir\Release\”。
     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045     2>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ZERO_CHECK.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 2 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(3) (Clean 个目标)。
     3>CoreClean:
         正在创建目录“freetype.dir\Release\”。
     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045ERO_CHECK.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 2 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(3) (Clean 个目标)。
     3>CoreClean:
         正在创建目录“freetype.dir\Release\”。
     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 2 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(3) (Clean 个目标)。
     3>CoreClean:
         正在创建目录“freetype.dir\Release\”。
     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(3) (Clean 个目标)。
     3>CoreClean:
         正在创建目录“freetype.dir\Release\”。
     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045     3>CoreClean:
         正在创建目录“freetype.dir\Release\”。
     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045     3>已完成生成项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\freetype.vcxproj”(Clean 个目标)的操作。
     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045     1>项目“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ALL_BUILD.vcxproj”(1)正在节点 1 上生成“C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\source\build_7045c03f\ZERO_CHECK.vcxproj”(2:3) (默认目标)。
c03f\ZERO_CHECK.vcxproj”(2:3) (默认目标)。
     2>PrepareForBuild:
     2>PrepareForBuild:
         已启用结构化输出。编译器诊断的格式设置将反映错误层次结构。有关详细信息，请参阅 https://aka.ms/cpp/structured-output。
utput。
         正在创建目录“x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\”。
       InitializeBuildStatus:
if you want to get more verbose errors, please see:
if you want to get more verbose errors, please see:
  -> C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\installdir.failed\logs\install.txt      
  -> C:\Users\A1387\AppData\Local\.xmake\cache\packages\2410\f\freetype\2.13.1\installdir.failed\logs\install.txt      
  => download https://github.com/miurahr/aqtinstall/releases/download/v3.1.17/aqt_x64.exe .. failed

we can also download these packages manually:
  - https://github.com/miurahr/aqtinstall/releases/download/v3.1.17/aqt_x64.exe
  - https://github.com/miurahr/aqtinstall/releases/download/v3.1.17/aqt_x64.exe
to the local search directories:
  - aqt_x64.exe, aqt-3.1.17
and we can run `xmake g --pkg_searchdirs=/xxx` to set the search directories.
  - aqt_x64.exe, aqt-3.1.17
and we can run `xmake g --pkg_searchdirs=/xxx` to set the search directories.
and we can run `xmake g --pkg_searchdirs=/xxx` to set the search directories.
  => download https://github.com/xmake-mirror/build-artifacts/releases/download/boost-1.86.0/windows-x64-vc14.40.33807-f1b18bc4ca114f2c9ac8606d97867946.7z .. failed

we can also download these packages manually:
  - https://github.com/xmake-mirror/build-artifacts/releases/download/boost-1.86.0/windows-x64-vc14.40.33807-f1b18bc4ca114f2c9ac8606d97867946.7z
to the local search directories:
to the local search directories:
  - boost-1.86.0.7z, windows-x64-vc14.40.33807-f1b18bc4ca114f2c9ac8606d97867946.7z
  - boost-1.86.0.7z, windows-x64-vc14.40.33807-f1b18bc4ca114f2c9ac8606d97867946.7z
  - boost-1.86.0.7z, windows-x64-vc14.40.33807-f1b18bc4ca114f2c9ac8606d97867946.7z
and we can run `xmake g --pkg_searchdirs=/xxx` to set the search directories.

**Comment by Mq-b** [comment]
无法安装 freetype 2.13.1

**Review by Mq-b — APPROVED** [review]

**Review by Mq-b — COMMENTED** [review]
no

**Review by Mq-b — CHANGES_REQUESTED** [review]
如上

**Comment by Arteiimis** [comment]
> 无法安装 freetype 2.13.1

xmake -v看一下完整日志

**Comment by Mq-b** [comment]
> > 无法安装 freetype 2.13.1
> 
> xmake -v看一下完整日志

```powershell
PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial\code\ModernCpp-ConcurrentProgramming-Tutorial> xmake -v
checking for platform ... windows
checking for architecture ... x64
checking for cl.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\cl.exe
checking for Microsoft Visual Studio (x64) version ... 2022
checking for Microsoft C/C++ Compiler (x64) version ... 19.41.34120
checking for zig ... no
checking for zig ... no
checking for nim ... no
checking for nim ... no
checking for git ... ok
checking for gzip ... ok
checking for tar ... ok
git rev-parse HEAD
checking for cmake ... no
checking for cmake ... no
checking for cmake ... no
checking for cmake ... ok
checking for pkgconf ... no
checking for pkgconf ... ok
checking for xmake::zlib ... no
checking for pkg-config ... ok
checking for zlib ... pkgconfig::zlib 1.3.1
checking for xmake::freetype ... no
checking for cl.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\cl.exe
checking for the c compiler (cc) ... cl.exe
checking for cl.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\cl.exe
checking for the c++ compiler (cxx) ... cl.exe
checking for link.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\link.exe
checking for the linker (ld) ... link.exe
> checking for c links(freetype)
> checking for c snippet(find_package/freetype)
checking for freetype ... no
checking for xmake::libogg ... libogg v1.3.4
checking for xmake::libflac ... libflac 1.4.3
checking for xmake::libvorbis ... libvorbis 1.3.7
checking for xmake::openal-soft ... openal-soft 1.23.1
checking for xmake::sfml ... no
> checking for c links(sfml)
> checking for c snippet(find_package/sfml)
checking for sfml ... no
checking for xmake::fmt ... fmt 11.0.2
checking for aqt ... no
checking for aqt ... no
checking for qmake ... ok
checking for Qt SDK directory ... D:/Qt/6.7.0/msvc2019_64
checking for Qt SDK version ... 6.7.0
checking for xmake::qt6base ... no
> checking for c links(qt6base)
> checking for c snippet(find_package/qt6base)
checking for qt6base ... no
checking for xmake::qt6core ... no
> checking for c links(qt6core)
> checking for c snippet(find_package/qt6core)
checking for qt6core ... no
checking for xmake::boost ... no
> checking for c links(boost)
> checking for c snippet(find_package/boost)
checking for boost ... no
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> freetype 2.13.1 [runtimes:"MD", from:sfml]
  -> sfml 2.6.0 [runtimes:"MD"]
  -> aqt 3.1.17 [from:qt6base]
  -> qt6base 6.8.0 [from:qt6core]
  -> qt6core 6.8.0 [runtimes:"MD"]
  -> boost 1.86.0 [runtimes:"MD"]
```

**Comment by Arteiimis** [comment]
> > > 无法安装 freetype 2.13.1
> > 
> > 
> > xmake -v看一下完整日志
> 
> ```powershell
> PS D:\project\ModernCpp-ConcurrentProgramming-Tutorial\code\ModernCpp-ConcurrentProgramming-Tutorial> xmake -v
> checking for platform ... windows
> checking for architecture ... x64
> checking for cl.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\cl.exe
> checking for Microsoft Visual Studio (x64) version ... 2022
> checking for Microsoft C/C++ Compiler (x64) version ... 19.41.34120
> checking for zig ... no
> checking for zig ... no
> checking for nim ... no
> checking for nim ... no
> checking for git ... ok
> checking for gzip ... ok
> checking for tar ... ok
> git rev-parse HEAD
> checking for cmake ... no
> checking for cmake ... no
> checking for cmake ... no
> checking for cmake ... ok
> checking for pkgconf ... no
> checking for pkgconf ... ok
> checking for xmake::zlib ... no
> checking for pkg-config ... ok
> checking for zlib ... pkgconfig::zlib 1.3.1
> checking for xmake::freetype ... no
> checking for cl.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\cl.exe
> checking for the c compiler (cc) ... cl.exe
> checking for cl.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\cl.exe
> checking for the c++ compiler (cxx) ... cl.exe
> checking for link.exe ... D:\visual studio 2022\VC\Tools\MSVC\14.41.34120\bin\HostX64\x64\link.exe
> checking for the linker (ld) ... link.exe
> > checking for c links(freetype)
> > checking for c snippet(find_package/freetype)
> checking for freetype ... no
> checking for xmake::libogg ... libogg v1.3.4
> checking for xmake::libflac ... libflac 1.4.3
> checking for xmake::libvorbis ... libvorbis 1.3.7
> checking for xmake::openal-soft ... openal-soft 1.23.1
> checking for xmake::sfml ... no
> > checking for c links(sfml)
> > checking for c snippet(find_package/sfml)
> checking for sfml ... no
> checking for xmake::fmt ... fmt 11.0.2
> checking for aqt ... no
> checking for aqt ... no
> checking for qmake ... ok
> checking for Qt SDK directory ... D:/Qt/6.7.0/msvc2019_64
> checking for Qt SDK version ... 6.7.0
> checking for xmake::qt6base ... no
> > checking for c links(qt6base)
> > checking for c snippet(find_package/qt6base)
> checking for qt6base ... no
> checking for xmake::qt6core ... no
> > checking for c links(qt6core)
> > checking for c snippet(find_package/qt6core)
> checking for qt6core ... no
> checking for xmake::boost ... no
> > checking for c links(boost)
> > checking for c snippet(find_package/boost)
> checking for boost ... no
> note: install or modify (m) these packages (pass -y to skip confirm)?
> in xmake-repo:
>   -> freetype 2.13.1 [runtimes:"MD", from:sfml]
>   -> sfml 2.6.0 [runtimes:"MD"]
>   -> aqt 3.1.17 [from:qt6base]
>   -> qt6base 6.8.0 [from:qt6core]
>   -> qt6core 6.8.0 [runtimes:"MD"]
>   -> boost 1.86.0 [runtimes:"MD"]
> ```

这个不是完整日志, 把报错也发出来, 太长可以发文件

**Pull request closed without merging**